libs/wingsdk/src/core/plugin-manager.ts

```diff
@@ -0,0 +1,145 @@
+import { readFileSync, writeFileSync } from "fs";
+import { resolve } from "path";
+import * as vm from "vm";
+import { IConstruct } from "constructs";
+
+/**
+ * Behavioral interface of a plugin hook.
+ */
+export interface ICompilationHook {
+  /**
+   *  name of plugin can be set by plugin as export
+   *
+   * @default - absolute path to plugin file
+   */
+  name: string;
+  /** preSynth hook */
+  preSynth?(app: IConstruct): void;
+  /** postSynth hook */
+  postSynth?(config: any): any;
+  /** validate hook */
+  validate?(config: any): void;
+}
+
+enum CompilationPhase {
+  PRE_SYNTH = "preSynth",
+  POST_SYNTH = "postSynth",
+  VALIDATE = "validate",
+}
+
+/**
+ * Plugin manager is responsible for loading hooks from plugins and
+ * managing their execution during compilation process.
+ */
+export class PluginManager {
+  private hooks: ICompilationHook[];
+
+  constructor(plugins: string[]) {
+    this.hooks = [];
+
+    for (const plugin of plugins) {
+      this.add(plugin);
+    }
+  }
+
+  /**
+   * Add plugins to the plugin manager based on the plugin type.
+   *
+   * @param pluginAbsolutePath the plugin to add
+   */
+  public add(pluginAbsolutePath: string) {
+    // maybe we support other plugin types in the future (e.g. npm modules)
+    if (!pluginAbsolutePath.endsWith(".js")) {
+      throw new Error(
+        `Currently only javascript files are supported as plugins. Got: ${pluginAbsolutePath}`
+      );
+    }
+
+    const hooks: ICompilationHook = {
+      name: pluginAbsolutePath,
+    };
+
+    const pluginDir = resolve(pluginAbsolutePath)
+      .split("/")
+      .slice(0, -1)
+      .join("/");
```

Comment:
If pluginAbsolutePath is absolute do we really need to call `resolve` on it?

If the function expects an absolute path and gets a non-absolute path then it would probably be better to throw / fail fast

Also we can use a helper from the `path ` module:
```suggestion
    const pluginDir = dirname(pluginAbsolutePath)
```